Data_Manipulation/voter_participation_crosswalk.R

```diff
@@ -0,0 +1,229 @@
+                             geometry = TRUE) %>%
+    mutate(pop10 = value) %>%
+    rename(block_id = GEOID) %>%
+    select(block_id,pop10,geometry) %>%
+    st_transform(26915) %>%
+    st_make_valid()
+  if(all(st_is_valid(blocks_sf))!= "TRUE") warning("The block level population file for the state contains invalid geometries")
+  return(blocks_sf)
+}
+# Function imports the ZCTAs level census voting population data 
+import_voting_population <- function(state_abv){
+  over_18_pop <-get_acs(geography ="zcta",
+                        variables = "DP05_0021",
+                        state = toupper(state_abv),
+                        year = 2019,
+                        geometry = TRUE) %>%
+    as.data.frame()%>%
+    rename(pop19 = estimate,
+           zcta = GEOID) %>%
+    select(zcta,pop19)
+}
+
+############ Function crosswalks ############ 
+# function generates a crosswalk from the precinct level results to the zcta level boundaries 
+crosswalk_votes_to_zctas <- function(raw_data_folder,data_folder, state_abv){
+  #### Step 1: Read in data  #### 
+  # precinct vote level data
+  vote_sf <- import_precinct_vote_data(state_abv, raw_data_folder)
+  
+  # zctas shape file 
+  zctas_sf <- import_zctas(state_abv)
+  
+  # block level census population data 
+  blocks_sf <- import_block_census(state_abv)
+  
+  # zcta level census population data 
+  over_18_pop <- import_voting_population(state_abv)
+  
+  #### Step 2: Return all intersections between the precinct boundaries and the ZCTA boundaries  #### 
+  # create intersections 
+  zctas_overlapping_polys <- st_buffer(st_intersection(zctas_sf,vote_sf),dist=0) %>%
+    mutate(precinct_zctas =  paste(geo_id, zcta, sep="_")) %>%
+    select(precinct_zctas,geo_id, zcta,geometry)
+  
+  
+  if(all(st_is_valid(zctas_overlapping_polys))!= "TRUE") warning("The zctas_sf contains invalid geometries")
+  if(length(unique(zctas_overlapping_polys$zcta)) != length(unique(zctas_sf$zcta))) warning("Some ZCTAs were dropped when creating the intersecting polygons") 
+  
+  #### Step 3: Attribute populations to each of the intersections by areal interpolation using block level census data  #### 
+  pop_of_intersection <- aw_interpolate(zctas_overlapping_polys,
+                                        tid = precinct_zctas,
+                                        source =  blocks_sf,
+                                        sid = block_id,
+                                        weight = "sum",
+                                        output = "sf",
+                                        extensive = "pop10") %>%
+    mutate(pop10= round(pop10,0))  %>%
+    sf::st_make_valid() %>%
+    na.omit()
+  if(sum(pop_of_intersection$pop10) != sum(blocks_sf$pop10)) warning(paste0("Not all the population has been attributed to the state. ",sum(pop_of_intersection$pop10), " is the sum of the population that has been attributed and ", sum(blocks_sf$pop10)," is the total population from the census file"))
```

Comment:
Throughout, can we rewrite the if-warning statements to be broken up over multiple lines? i.e.:

```r
if (thing){
  warning()
}
```